### ISSUE TITLE
Export of list of results

### ISSUE THREAD

**Issue by peterwebster** [ukwa/shine#10]
For any query, users need to be able to export as CSV the list of results, one line per result.

Exact composition to be define, but I imagine two options:  Full and Brief

Full = all metadata associated with a resource, including facet values [which may not be straightforward if there are multiple values.]

Brief = title, crawl date, URL

**Comment by peterwebster** [comment]
Thanks @kinmanli : I successfully downloaded both Brief and Full for a result set of 35 results, in a few seconds in each case:
http://www.webarchive.org.uk/shine/search/advanced?query=%22biblical+creation+society%22&action=search&hostDomainPublicSuffix=grace.org.uk&sort=content_type_norm&order=asc

but a result set of 1165 results took more than three minutes (before I stopped it):

http://www.webarchive.org.uk/shine/search/advanced?page=1&facet.fields=public_suffix&addFacet=&action=search&query=%22biblical+creation+society%22&sort=content_type_norm&order=asc

Even with the warning message that the download may take a while, this may need looking at eventually.

**Comment by peterwebster** [comment]
Also [small point]:  once the download is complete, the view ought to refresh to remove the 'Exporting to CSV..' message, and show (in the same position) the message 'Download complete.'  @kinmanli

**Comment by peterwebster** [comment]
[at 10:05, I got this - after about 8 minutes]

Proxy Error

The proxy server received an invalid response from an upstream server.
The proxy server could not handle the request GET /shine/search/export.

Reason: Error reading from remote server

Apache/2.2.3 (Red Hat) Server at www.webarchive.org.uk Port 80

**Comment by peterwebster** [comment]
.... and on inspection, the two CSV files I did get out were empty. @kinmanli

**Comment by peterwebster** [comment]
Hi @kinmanli :  I've just tested this again, and the CSV option responds as I would expect visually, Chrome shows me the file that is downloaded, but the file has no data in it, when I open it up.

**Comment by kinmanli** [comment]
@peterwebster What were the parameters in the URL?

**Comment by peterwebster** [comment]
http://www.webarchive.org.uk/shine/search/advanced?query=%22biblical+creation+society%22&action=search&hostDomainPublicSuffix=ac.uk&sort=content_type_norm&order=asc

**Comment by kinmanli** [comment]
@peterwebster this only happening in "advanced search"?

**Comment by peterwebster** [comment]
@kinmanli it seems so - just got Brief and Full successfully from the simple search

**Comment by kinmanli** [comment]
@peterwebster it seems to be anything that has a facet (except "General Content Type")

**Comment by kinmanli** [comment]
@peterwebster fixed

**Comment by peterwebster** [comment]
@kinmanli Just tested this, and it seems to be OK. I did make an unsuccessful attempt before that, where the browser sat waiting for a response. Think this may be while Solr was waking up ?

**Comment by peterwebster** [comment]
[copied from earlier in the ticket, two small outstanding enhancements, now lower priority]

[not a high priority] As discussed today, two enhancements to the CSV output:

(i) output should include the field names as column headings;

(ii) output should also include some basic information about the search that generated the output:

'Results from the British Library's Shine interface, (webarchive.org.uk/shine), on [dd/mm/yyyy]. 
[new line] Search Summary: [lines as printed in the UI search summary, separated by semi-colons]'

eg.
Results from the British Library's Shine interface (webarchive.org.uk/shine), on 17/09/2014.
Search Summary: Search Terms: wikipedia; None of these words: essay

@kinmanli

**Comment by kinmanli** [comment]
![2015-03-11 13-39-41_microsoft excel - export 134](https://cloud.githubusercontent.com/assets/1652105/6597381/1f86ef68-c7f4-11e4-8165-54c309f42539.png)

![2015-03-11 13-46-14_microsoft excel - export 136](https://cloud.githubusercontent.com/assets/1652105/6597503/08dccad4-c7f5-11e4-9ff7-d4e12a5a5815.png)

**Comment by tokee** [comment]
Internally in Solr, direct paging to large page numbers is discouraged as the search time grows fast (there's an internal priority queue of size page*pagesize). A quick look in Shine.java indicates that this is done when doing a CSV-export.

For export, the "old" way is to use cursorMark (https://cwiki.apache.org/confluence/display/solr/Pagination+of+Results) and the new way is to use streaming export (https://cwiki.apache.org/confluence/display/solr/Exporting+Result+Sets). However, the new way requires Solr 4.10 and that the export fields are DocValues, so pagination with cursorMark seem to be the conservative choise.

**Comment by peterwebster** [comment]
The results shown in the screenshots look good, but when I tried to test just now,, it fails:


http://www.webarchive.org.uk/shine/search/export?proximityPhrase1=rowan+williams&proximityPhrase2=sharia&proximity=25&action=search&dateStart=2008-02-01&dateEnd=2008-12-31&sort=content_type_norm&order=asc&exportType=csv&version=full
@kinmanli

**Comment by peterwebster** [comment]
After discussion, agreed that we should go with Toke's conservative option @kinmanli

**Comment by peterwebster** [comment]
Hi @kinmanli : the 'Results from the ..' and 'Search Summary...' lines seem to have dropped out of these CSV outputs (?)

**Comment by kinmanli** [comment]
@peterwebster advanced search details was missing. Now working.

![2015-03-30 15-14-02_microsoft excel - export 4](https://cloud.githubusercontent.com/assets/1652105/6898038/6e219d12-d6ef-11e4-9e0e-b33fc71b9321.png)
